## This is a simple Image classification task using MNIST dataset

In [1]:
# importing the MNIST dataset
from sklearn.datasets import fetch_openml
# fetch_* functions load reaal-life datasets
# load_* fucntons load small toy datasets
# make_* functions make fake datasets
# The generated datasets are returend as (X, y) tuples
# The generated datasets contain input data and targets as NumPy arrays
# Other datasets are returned as sklearn.utils.Bunch objects
# These datasets are dictionaries whose entries can be accessed
# The attirbutes for access are: "DESCR", "data", "target"
# DESCR: returns the description of the dataset
# data: The input data, usually as a 2D NumPy array
# target: The labels, usually as 1D numpy array
mnist = fetch_openml('mnist_784', as_frame=False)
# fetch_openml() returns data as pandas df and labels as pandas series
# This is unusual for MNIST dataset which contains images and isnt ideal
# To get data as NumPy arrays instead we set "as_frame=False" 

/home/umarsajjad/anaconda3/envs/sklearn_env/lib/python3.10/site-packages/sklearn/datasets/_openml.py:932: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [2]:
# Let's store the data into variables for ease of access
X, y = mnist.data, mnist.target
X

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [3]:
# Let's take a look at the shapes of the X (inputs) and y (labels)
print(X.shape)
print(y.shape)

(70000, 784)
(70000,)


* As can seen by the shape of the inputs, there are a total of 70,000 images and each image has 784 features (num pixels-28x28 pixel image)
* Each feature can have a value between 0(white) and 255(black)

In [ ]:
# Let's take peak at one example from the dataset
# Let's make function that plots the digit
import matplotlib.pyplot as plt
# we are going to use imshow() function of matplotlib to display the image
# we use cmap="binary" to get a grayscale color map (0=white, 255=black)

def plot_digit(image_data):